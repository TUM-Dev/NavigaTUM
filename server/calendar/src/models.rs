use chrono::NaiveDateTime;
use diesel::prelude::*;
use diesel::Insertable;

#[derive(Insertable, Queryable, AsChangeset)]
#[diesel(table_name = crate::schema::calendar)]
pub struct XMLEvent {
    pub key: String,
    pub tumonline_id: i32,
    pub dtstart: NaiveDateTime,
    pub dtend: NaiveDateTime,
    pub dtstamp: NaiveDateTime,
    pub event_id: i32,
    pub event_title: String,
    pub single_event_id: i32,
    pub single_event_type_id: String,
    pub single_event_type_name: String,
    pub event_type_id: String,
    pub event_type_name: Option<String>,
    pub course_type_name: Option<String>,
    pub course_type: Option<String>,
    pub course_code: Option<String>,
    pub course_semester_hours: Option<i32>,
    pub group_id: Option<String>,
    pub xgroup: Option<String>,
    pub status_id: String,
    pub status: String,
    pub comment: String,
    pub last_scrape: NaiveDateTime,
}
