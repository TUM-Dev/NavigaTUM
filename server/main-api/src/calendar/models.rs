use crate::models::Location;
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use std::collections::HashMap;

#[derive(Serialize, Deserialize, Clone, Debug)]
pub(super) struct EventsCollection {
    pub(super) events: HashMap<String, LocationEvents>,
    pub(super) max_last_sync: DateTime<Utc>,
}

#[derive(Serialize, Deserialize, Clone, Debug)]
pub(super) struct CalendarLocation {
    pub key: String,
    pub name: String,
    pub last_calendar_scrape_at: Option<DateTime<Utc>>,
    pub calendar_url: Option<String>,
    pub type_common_name: String,
    pub r#type: String,
}

impl From<Location> for CalendarLocation {
    fn from(loc: Location) -> Self {
        Self {
            key: loc.key,
            name: loc.name,
            last_calendar_scrape_at: loc.last_calendar_scrape_at,
            calendar_url: loc.calendar_url,
            type_common_name: loc.type_common_name,
            r#type: loc.r#type,
        }
    }
}

#[derive(Serialize, Deserialize, Clone, Debug)]
pub(super) struct LocationEvents {
    pub(super) events: Vec<Event>,
    pub(super) location: CalendarLocation,
}

#[derive(Serialize, Deserialize, Clone, Debug, sqlx::Type)]
pub(super) struct Event {
    pub(super) id: i32,
    /// e.g. 5121.EG.003
    pub(super) room_code: String,
    /// e.g. 2018-01-01T00:00:00
    pub(super) start_at: DateTime<Utc>,
    /// e.g. 2019-01-01T00:00:00
    pub(super) end_at: DateTime<Utc>,
    /// e.g. Quantenteleportation
    pub(super) stp_title_de: String,
    /// e.g. Quantum teleportation
    pub(super) stp_title_en: String,
    /// e.g. Vorlesung mit Zentral√ºbung
    pub(super) stp_type: String,
    /// e.g. lecture
    pub(super) entry_type: EventType,
    /// e.g. Abhaltung
    pub(super) detailed_entry_type: String,
}

impl Event {
    pub async fn store(
        &self,
        tx: &mut sqlx::Transaction<'_, sqlx::Postgres>,
    ) -> Result<sqlx::postgres::PgQueryResult, sqlx::Error> {
        sqlx::query!(
            r#"INSERT INTO calendar (id,room_code,start_at,end_at,stp_title_de,stp_title_en,stp_type,entry_type,detailed_entry_type)
            VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9)
            ON CONFLICT (id) DO UPDATE SET
             room_code = EXCLUDED.room_code,
             start_at = EXCLUDED.start_at,
             end_at = EXCLUDED.end_at,
             stp_title_de = EXCLUDED.stp_title_de,
             stp_title_en = EXCLUDED.stp_title_en,
             stp_type = EXCLUDED.stp_type,
             entry_type = EXCLUDED.entry_type,
             detailed_entry_type = EXCLUDED.detailed_entry_type"#,
            self.id,
            self.room_code,
            self.start_at,
            self.end_at,
            self.stp_title_de,
            self.stp_title_en,
            self.stp_type,
            self.entry_type.clone() as EventType, // see https://github.com/launchbadge/sqlx/issues/1004 => our type is not possible (?)
            self.detailed_entry_type,
        ).execute(&mut **tx).await
    }
}

#[derive(Serialize, Deserialize, Clone, Debug, sqlx::Type)]
#[sqlx(type_name = "EventType")]
#[sqlx(rename_all = "lowercase")]
#[serde(rename_all = "lowercase")]
pub enum EventType {
    Lecture,
    Exercise,
    Exam,
    Barred,
    Other,
}
