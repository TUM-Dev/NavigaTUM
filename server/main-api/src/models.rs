use diesel::prelude::*;

#[derive(Queryable, Debug, Clone)]
pub struct DBRoomEntry {
    pub key: String,
    pub name: String,
    pub tumonline_room_nr: Option<i32>,
    pub arch_name: Option<String>,
    pub type_: String,
    pub type_common_name: String,
    pub lat: f32,
    pub lon: f32,
    pub data: String,
}
