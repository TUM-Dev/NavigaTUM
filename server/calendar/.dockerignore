target
Cargo.lock
data
.hypothesis
