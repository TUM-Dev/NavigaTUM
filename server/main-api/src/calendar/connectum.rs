use std::time::{Duration, Instant};
use std::{env, io};

use chrono::{DateTime, Utc};
use log::{debug, error, warn};
use oauth2::basic::{BasicClient, BasicTokenResponse};
use oauth2::reqwest::async_http_client;
use oauth2::url::Url;
use oauth2::{AuthUrl, ClientId, ClientSecret, Scope, TokenResponse, TokenUrl};
use sqlx::PgPool;

use crate::calendar::models::Event;

pub(in crate::calendar) struct APIRequestor {
    client: reqwest::Client,
    pool: PgPool,
    oauth_token: Option<(Instant, BasicTokenResponse)>,
}

impl From<&PgPool> for APIRequestor {
    fn from(pool: &PgPool) -> Self {
        let keep_alive = Duration::from_secs(2);
        let client = reqwest::Client::builder()
            .tcp_keepalive(keep_alive)
            .http2_keep_alive_while_idle(true)
            .http2_keep_alive_interval(keep_alive)
            .gzip(true)
            .zstd(true)
            .brotli(true)
            .deflate(true)
            .build()
            .expect("the request client builder is correctly configured");
        Self {
            client,
            pool: pool.clone(),
            oauth_token: None,
        }
    }
}

impl APIRequestor {
    pub(crate) async fn refresh(&mut self, id: &str) -> Result<(), crate::BoxedError> {
        let sync_start = Utc::now();
        let token = self.try_unwrap_or_refresh_token().await?;
        let start = Instant::now();
        let url = format!("https://campus.tum.de/tumonline/co/connectum/api/rooms/{id}/calendars");
        let events: Vec<Event> = self
            .client
            .get(url)
            .bearer_auth(token)
            .send()
            .await?
            .json()
            .await?;
        debug!(
            "finished fetching for {cnt} calendar events of {id} in {elapsed:?}",
            cnt = events.len(),
            elapsed = start.elapsed()
        );
        let events = events
            .into_iter()
            .map(|mut e| {
                e.room_code = id.into();
                e
            })
            .collect::<Vec<Event>>();
        self.store(&events, &sync_start, id).await?;
        Ok(())
    }
    fn should_refresh_token(&self) -> bool {
        if let Some((start, token)) = &self.oauth_token {
            if let Some(expires_in) = token.expires_in() {
                return expires_in - start.elapsed() < Duration::from_secs(10);
            }
        }
        true
    }
    async fn try_unwrap_or_refresh_token(&mut self) -> Result<String, crate::BoxedError> {
        if self.should_refresh_token() {
            self.oauth_token = Some(Self::fetch_new_oauth_token().await?);
        }

        let at = self
            .oauth_token
            .as_ref()
            .expect("the token has been set in the last step")
            .1
            .access_token();
        Ok(at.secret().clone())
    }
}

impl APIRequestor {
    async fn store(
        &self,
        events: &[Event],
        last_calendar_scrape_at: &DateTime<Utc>,
        id: &str,
    ) -> Result<(), crate::BoxedError> {
        let start = Instant::now();
        // insert into db
        let mut tx = self.pool.begin().await?;
        if let Err(e) = self.delete_events(&mut tx, id).await {
            error!("could not delete existing events because {e:?}");
            tx.rollback().await?;
            return Err(e.into());
        }
        let mut failed: Option<(usize, sqlx::Error)> = None;
        for event in events.iter() {
            // conflicts cannot occur because all values for said room were dropped
            if let Err(e) = event.store(&mut tx).await {
                failed = match failed {
                    Some((i, e0)) => Some((i + 1, e0)),
                    None => Some((1, e)),
                };
            }
        }
        if let Some((cnt, e)) = failed {
            warn!(
                "{cnt}/{total} events could not be inserted because of {e:?}",
                total = events.len()
            );
        }
        if let Err(e) = self
            .update_last_calendar_scrape_at(&mut tx, id, last_calendar_scrape_at)
            .await
        {
            error!("could not update last_calendar_scrape_at because {e:?}");
            tx.rollback().await?;
            return Err(e.into());
        }
        tx.commit().await?;
        debug!(
            "finished inserting into the db for {id} in {elapsed:?}",
            elapsed = start.elapsed()
        );
        Ok(())
    }

    async fn fetch_new_oauth_token() -> Result<(Instant, BasicTokenResponse), crate::BoxedError> {
        let client_id = env::var("CONNECTUM_OAUTH_CLIENT_ID")
            .map_err(|e| {
                error!("CONNECTUM_OAUTH_CLIENT_ID needs to be set: {e:?}");
                io::Error::other("please configure the environment variable CONNECTUM_OAUTH_CLIENT_ID to use this endpoint")
            })?
            .trim().into();
        let client_secret = env::var("CONNECTUM_OAUTH_CLIENT_SECRET")
            .map_err(|e| {
                error!("CONNECTUM_OAUTH_CLIENT_SECRET needs to be set: {e:?}");
                io::Error::other("please configure the environment variable CONNECTUM_OAUTH_CLIENT_SECRET to use this endpoint")
            })?
            .trim().into();

        // for urls see https://campus.tum.de/tumonline/co/public/sec/auth/realms/CAMPUSonline/.well-known/openid-configuration
        let auth_url = Url::parse("https://campus.tum.de/tumonline/co/public/sec/auth/realms/CAMPUSonline/protocol/openid-connect/auth")?;
        let token_url = Url::parse("https://campus.tum.de/tumonline/co/public/sec/auth/realms/CAMPUSonline/protocol/openid-connect/token")?;

        let token = BasicClient::new(
            ClientId::new(client_id),
            Some(ClientSecret::new(client_secret)),
            AuthUrl::from_url(auth_url),
            Some(TokenUrl::from_url(token_url)),
        )
        .exchange_client_credentials()
        .add_scope(Scope::new("connectum-rooms.read".into()))
        .request_async(async_http_client)
        .await;
        Ok((Instant::now(), token?))
    }
    async fn delete_events(
        &self,
        tx: &mut sqlx::Transaction<'_, sqlx::Postgres>,
        id: &str,
    ) -> Result<sqlx::postgres::PgQueryResult, sqlx::Error> {
        sqlx::query!(r#"DELETE FROM calendar WHERE room_code = $1"#, id)
            .execute(&mut **tx)
            .await
    }
    async fn update_last_calendar_scrape_at(
        &self,
        tx: &mut sqlx::Transaction<'_, sqlx::Postgres>,
        id: &str,
        last_calendar_scrape_at: &DateTime<Utc>,
    ) -> Result<sqlx::postgres::PgQueryResult, sqlx::Error> {
        sqlx::query!(
            "UPDATE en SET last_calendar_scrape_at = $1 WHERE key=$2",
            last_calendar_scrape_at,
            id
        )
        .execute(&mut **tx)
        .await?;
        sqlx::query!(
            "UPDATE de SET last_calendar_scrape_at = $1 WHERE key=$2",
            last_calendar_scrape_at,
            id
        )
        .execute(&mut **tx)
        .await
    }
}
