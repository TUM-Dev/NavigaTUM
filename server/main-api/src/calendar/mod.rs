use std::collections::HashMap;

use actix_web::{post, web, HttpResponse};
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use sqlx::PgPool;
use tracing::error;

use crate::calendar::models::{CalendarLocation, Event, LocationEvents};
use crate::limited::hash_map::LimitedHashMap;
use crate::limited::vec::LimitedVec;

mod connectum;
mod models;
pub mod refresh;

#[derive(Serialize, Deserialize, Clone, Debug)]
pub struct Arguments {
    ids: Vec<String>,
    /// eg. 2039-01-19T03:14:07+1
    start_after: DateTime<Utc>,
    /// eg. 2042-01-07T00:00:00 UTC
    end_before: DateTime<Utc>,
}

impl Arguments {
    fn validate_ids(&self) -> Result<Vec<String>, HttpResponse> {
        let ids = self
            .ids
            .clone()
            .into_iter()
            .map(|s| s.replace(|c: char| c.is_whitespace() || c.is_control(), ""))
            .collect::<Vec<String>>();
        if ids.len() > 10 {
            return Err(HttpResponse::BadRequest().body("Too many ids to query. We suspect that users don't need this. If you need this limit increased, please send us a message"));
        };
        if ids.is_empty() {
            return Err(HttpResponse::BadRequest().body("No id requested"));
        };
        Ok(ids)
    }
}

#[post("/api/calendar")]
pub async fn calendar_handler(
    web::Json(args): web::Json<Arguments>,
    data: web::Data<crate::AppData>,
) -> HttpResponse {
    let ids = match args.validate_ids() {
        Ok(ids) => ids,
        Err(e) => return e,
    };
    let locations = match get_locations(&data.pool, &ids).await {
        Ok(l) => l.0,
        Err(e) => return e,
    };
    if let Err(e) = validate_locations(&ids, &locations) {
        return e;
    }
    match get_from_db(&data.pool, &locations, &args.start_after, &args.end_before).await {
        Ok(events) => HttpResponse::Ok().json(events),
        Err(e) => {
            error!("could not get entries from the db for {ids:?} because {e:?}");
            HttpResponse::InternalServerError()
                .body("could not get calendar entries, please try again later")
        }
    }
}

fn validate_locations(ids: &[String], locations: &[CalendarLocation]) -> Result<(), HttpResponse> {
    for id in ids {
        if !locations.iter().any(|l| &l.key == id) {
            return Err(HttpResponse::BadRequest().body("Requested id {id} does not exist"));
        }
    }
    assert_eq!(locations.len(), ids.len());
    for loc in locations {
        if loc.last_calendar_scrape_at.is_none() {
            return Err(HttpResponse::ServiceUnavailable().body(format!("Room {key}/{url:?} calendar entry is currently in the process of being scraped, please try again later", key = loc.key, url = loc.calendar_url)));
        };
    }
    for loc in locations {
        if loc.calendar_url.is_none() {
            return Err(HttpResponse::NotFound()
                .content_type("text/plain")
                .body(format!(
                    "Room {key}/{url:?} does not have a calendar",
                    key = loc.key,
                    url = loc.calendar_url
                )));
        };
    }
    Ok(())
}

#[tracing::instrument(skip(pool))]
async fn get_locations(
    pool: &PgPool,
    ids: &[String],
) -> Result<LimitedVec<CalendarLocation>, HttpResponse> {
    match sqlx::query_as!(CalendarLocation, "SELECT key,name,last_calendar_scrape_at,calendar_url,type,type_common_name FROM de WHERE key = ANY($1::text[])", ids).fetch_all(pool).await {
        Err(e) => {
            error!("could not refetch due to {e:?}");
            Err(HttpResponse::InternalServerError().body("could not get calendar entries, please try again later"))
        }
        Ok(locations) => Ok(LimitedVec(locations)),
    }
}

#[tracing::instrument(skip(pool),ret(level = tracing::Level::TRACE))]
async fn get_from_db(
    pool: &PgPool,
    locations: &[CalendarLocation],
    start_after: &DateTime<Utc>,
    end_before: &DateTime<Utc>,
) -> Result<LimitedHashMap<String, LocationEvents>, crate::BoxedError> {
    let mut located_events: HashMap<String, LocationEvents> = HashMap::new();
    for location in locations {
        let events = sqlx::query_as!(Event, r#"SELECT id,room_code,start_at,end_at,stp_title_de,stp_title_en,stp_type,entry_type,detailed_entry_type
            FROM calendar
            WHERE room_code = $1 AND start_at >= $2 AND end_at <= $3"#,
            location.key, start_after, end_before).fetch_all(pool).await?;
        located_events.insert(
            location.key.clone(),
            LocationEvents {
                location: location.clone(),
                events,
            },
        );
    }
    Ok(LimitedHashMap(located_events))
}

#[cfg(test)]
mod tests {
    use std::sync::Arc;

    use actix_web::http::header::ContentType;
    use actix_web::test;
    use actix_web::App;
    use lazy_static::lazy_static;
    use pretty_assertions::assert_eq;
    use serde_json::Value;

    use crate::setup::tests::PostgresTestContainer;
    use crate::AppData;

    use super::*;

    lazy_static! {
        static ref TIME_Y2K: DateTime<Utc> =
            DateTime::parse_from_rfc3339("2000-01-01T00:00:00-00:00")
                .unwrap()
                .to_utc();
        static ref TIME_2010: DateTime<Utc> =
            DateTime::parse_from_rfc3339("2010-01-01T00:00:00-00:00")
                .unwrap()
                .to_utc();
        static ref TIME_2012: DateTime<Utc> =
            DateTime::parse_from_rfc3339("2012-01-01T00:00:00-00:00")
                .unwrap()
                .to_utc();
        static ref TIME_2014: DateTime<Utc> =
            DateTime::parse_from_rfc3339("2014-01-01T00:00:00-00:00")
                .unwrap()
                .to_utc();
        static ref TIME_2016: DateTime<Utc> =
            DateTime::parse_from_rfc3339("2016-01-01T00:00:00-00:00")
                .unwrap()
                .to_utc();
        static ref TIME_2018: DateTime<Utc> =
            DateTime::parse_from_rfc3339("2018-01-01T00:00:00-00:00")
                .unwrap()
                .to_utc();
        static ref TIME_2020: DateTime<Utc> =
            DateTime::parse_from_rfc3339("2020-01-01T00:00:00-00:00")
                .unwrap()
                .to_utc();
    }
    fn sample_data() -> (Vec<(String, Value)>, Vec<Event>) {
        (
            vec![
                (
                    "5121.EG.003".into(),
                    serde_json::json!({"aliases":["003@5121"],"coords":{"accuracy":"building","lat":48.26842603718826,"lon":11.677995005953209,"source":"inferred"},"id":"5121.EG.003","maps":{"default":"interactive"},"name":"5121.EG.003 (Computerraum)","parent_names":["Standorte","Garching Forschungszentrum","Physik","Maier-Leibnitz-Laboratorium (MLL), TUM & LMU","Atlashalle"],"parents":["root","garching","physik","mll","5121"],"poi":{"nearby_public_transport":{"mvg":[]}},"props":{"calendar_url":"https://campus.tum.de/3","computed":[{"name":"Raumkennung","text":"5121.EG.003"},{"name":"Architekten-Name","text":"003"},{"name":"Stockwerk","text":"Erdgeschoss"},{"name":"Adresse","text":"Am Coulombwall 6, 85748 Garching b. München"}],"operator":{"code":"TUPELMU","id":39536,"name":"Ludwig-Maximilians-Universität München (LMU)","url":"https://campus.tum.de/tumonline/webnav.navigate_to?corg=39536"},"tumonline_room_nr":45064},"ranking_factors":{"rank_combined":10,"rank_type":100,"rank_usage":10},"sources":{"base":[{"name":"TUMonline","url":"https://campus.tum.de/tumonline/ee/ui/ca2/app/desktop/#/pl/ui/$ctx/45064"}]},"type":"room","type_common_name":"Serverraum","usage":{"din_277":"TF8.9","din_277_desc":"Sonstige betriebstechnische Anlagen","name":"Serverraum"},"redirect_url":"/room/5121.EG.003"}),
                ),
                (
                    "5121.EG.002".into(),
                    serde_json::json!({"aliases":["002@5121"],"coords":{"accuracy":"building","lat":48.26842603718826,"lon":11.677995005953209,"source":"inferred"},"id":"5121.EG.002","maps":{"default":"interactive"},"name":"5121.EG.002 (Testroom)","parent_names":["Standorte","Garching Forschungszentrum","Physik","Maier-Leibnitz-Laboratorium (MLL),TUM & LMU","Atlashalle"],"parents":["root","garching","physik","mll","5121"],"poi":{"nearby_public_transport":{"mvg":[]}},"props":{"computed":[{"name":"Raumkennung","text":"5121.EG.002"},{"name":"Architekten-Name","text":"002"},{"name":"Stockwerk","text":"Erdgeschoss"},{"name":"Adresse","text":"Am Coulombwall 6,85748 Garching b. München"}  ],"operator":{"code":"TUPELMU","id":39536,"name":"Ludwig-Maximilians-Universität München (LMU)","url":"https://campus.tum.de/tumonline/webnav.navigate_to?corg=39536"},"tumonline_room_nr":44904},"ranking_factors":{"rank_combined":10,"rank_type":100,"rank_usage":10},"sources":{"base":[{"name":"TUMonline","url":"https://campus.tum.de/tumonline/ee/ui/ca2/app/desktop/#/pl/ui/$ctx/44904"}  ]},"type":"room","type_common_name":"Versuchshalle","usage":{"din_277":"NF3.3","din_277_desc":"Technologische Labors","name":"Versuchshalle"},"redirect_url":"/room/5121.EG.002"}),
                ),
                (
                    "5121.EG.001".into(),
                    serde_json::json!({"aliases":["001@5121"],"coords":{"accuracy":"building","lat":48.26842603718826,"lon":11.677995005953209,"source":"inferred"},"id":"5121.EG.001","maps":{"default":"interactive"},"name":"5121.EG.001 (Montage- und Versuchshalle)","parent_names":["Standorte","Garching Forschungszentrum","Physik","Maier-Leibnitz-Laboratorium (MLL),TUM & LMU","Atlashalle"],"parents":["root","garching","physik","mll","5121"],"poi":{"nearby_public_transport":{"mvg":[]}},"props":{"calendar_url":"https://campus.tum.de/1","computed":[{"name":"Raumkennung","text":"5121.EG.001"},{"name":"Architekten-Name","text":"001"},{"name":"Stockwerk","text":"Erdgeschoss"},{"name":"Adresse","text":"Am Coulombwall 6,85748 Garching b. München"}  ],"operator":{"code":"TUPELMU","id":39536,"name":"Ludwig-Maximilians-Universität München (LMU)","url":"https://campus.tum.de/tumonline/webnav.navigate_to?corg=39536"},"tumonline_room_nr":44904},"ranking_factors":{"rank_combined":10,"rank_type":100,"rank_usage":10},"sources":{"base":[{"name":"TUMonline","url":"https://campus.tum.de/tumonline/ee/ui/ca2/app/desktop/#/pl/ui/$ctx/44904"}  ]},"type":"room","type_common_name":"Versuchshalle","usage":{"din_277":"NF3.3","din_277_desc":"Technologische Labors","name":"Versuchshalle"},"redirect_url":"/room/5121.EG.001"}),
                ),
            ],
            vec![
                Event {
                    id: 1,
                    room_code: "5121.EG.003".into(),
                    start_at: *TIME_2012,
                    end_at: *TIME_2014,
                    stp_title_de: "Quantenteleportation".into(),
                    stp_title_en: "Quantum teleportation".into(),
                    stp_type: "Vorlesung mit Zentralübung".into(),
                    entry_type: models::EventType::Lecture.to_string(),
                    detailed_entry_type: "Abhaltung".into(),
                },
                Event {
                    id: 2,
                    room_code: "5121.EG.003".into(),
                    start_at: *TIME_2014,
                    end_at: *TIME_2016,
                    stp_title_de: "Quantenteleportation 2".into(),
                    stp_title_en: "Quantum teleportation 2".into(),
                    stp_type: "Vorlesung mit Zentralübung".into(),
                    entry_type: models::EventType::Lecture.to_string(),
                    detailed_entry_type: "Abhaltung".into(),
                },
                Event {
                    id: 3,
                    room_code: "5121.EG.001".into(),
                    start_at: *TIME_2014,
                    end_at: *TIME_2016,
                    stp_title_de: "Wartung".into(),
                    stp_title_en: "maintenance".into(),
                    stp_type: "Vorlesung mit Zentralübung".into(),
                    entry_type: models::EventType::Barred.to_string(),
                    detailed_entry_type: "Abhaltung".into(),
                },
                Event {
                    id: 4,
                    room_code: "5121.EG.001".into(),
                    start_at: *TIME_Y2K,
                    end_at: *TIME_2020,
                    stp_title_de: "Quantenteleportation 3".into(),
                    stp_title_en: "Quantum teleportation 3".into(),
                    stp_type: "Vorlesung".into(),
                    entry_type: models::EventType::Other.to_string(),
                    detailed_entry_type: "Abhaltung".into(),
                },
                Event {
                    id: 5,
                    room_code: "5121.EG.001".into(),
                    start_at: *TIME_Y2K,
                    end_at: *TIME_2010,
                    stp_title_de: "Quantenteleportation 3".into(),
                    stp_title_en: "Quantum teleportation 3".into(),
                    stp_type: "Vorlesung".into(),
                    entry_type: models::EventType::Exam.to_string(),
                    detailed_entry_type: "Abhaltung".into(),
                },
            ],
        )
    }

    async fn load_sample_data(pool: &PgPool, now_rfc3339: &str) {
        let mut tx = pool.begin().await.unwrap();
        let (locations, events) = sample_data();
        for (key, data) in locations {
            for lang in ["de", "en"] {
                let query = format!("INSERT INTO {lang}(key,data,last_calendar_scrape_at) VALUES ('{key}','{data}','{now_rfc3339}')");
                sqlx::query(&query).execute(&mut *tx).await.unwrap();
            }
        }

        for event in events {
            event.store(&mut tx).await.unwrap();
        }
        tx.commit().await.unwrap();
    }

    #[actix_web::test]
    async fn test_index_get() {
        // setup + load data into postgres
        let pg = PostgresTestContainer::new().await;
        let now = Utc::now();
        let now = now.to_rfc3339_opts(chrono::SecondsFormat::Secs, true); // throwing away accuracy for simpler testing
        load_sample_data(&pg.pool, &now).await;
        // set up the http service/api/calendar
        let app = test::init_service(
            App::new()
                .app_data(web::Data::new(AppData {
                    pool: pg.pool.clone(),
                    meilisearch_initialised: Arc::new(Default::default()),
                }))
                .service(calendar_handler),
        )
        .await;
        // -- send requests and assert response --
        {
            // missing required query parameters
            let req = test::TestRequest::post()
                .uri("/api/calendar")
                .insert_header(ContentType::json())
                .to_request();
            let (_, resp) = test::call_service(&app, req).await.into_parts();
            run_testcase(
                resp,
                400,
                "Json deserialize error: EOF while parsing a value at line 1 column 0",
            )
            .await;
        }
        {
            // missing required query parameters
            let args = Arguments {
                end_before: Utc::now(),
                start_after: Utc::now(),
                ids: vec![],
            };
            let req = test::TestRequest::post()
                .uri("/api/calendar")
                .set_json(args)
                .insert_header(ContentType::json())
                .to_request();
            let (_, resp) = test::call_service(&app, req).await.into_parts();
            run_testcase(resp, 400, "No id requested").await;
        }
        {
            // way too many parameters
            let args = Arguments {
                end_before: Utc::now(),
                start_after: Utc::now(),
                ids: (0..10_000).map(|i| i.to_string()).collect(),
            };
            let req = test::TestRequest::post()
                .uri("/api/calendar")
                .set_json(args)
                .insert_header(ContentType::json())
                .to_request();
            let (_, resp) = test::call_service(&app, req).await.into_parts();
            run_testcase(resp, 400, "Too many ids to query. We suspect that users don't need this. If you need this limit increased, please send us a message").await;
        }
        {
            // room without a calendar
            let args = Arguments {
                end_before: Utc::now(),
                start_after: Utc::now(),
                ids: vec!["5121.EG.002".into()],
            };
            let req = test::TestRequest::post()
                .uri("/api/calendar")
                .set_json(args)
                .insert_header(ContentType::json())
                .to_request();
            let (_, resp) = test::call_service(&app, req).await.into_parts();
            run_testcase(resp, 404, "Room 5121.EG.002/None does not have a calendar").await;
        }
        {
            // show all entries of 5121.EG.003
            let args = Arguments {
                start_after: *TIME_Y2K,
                end_before: *TIME_2020,
                ids: vec!["5121.EG.003".into()],
            };
            let req = test::TestRequest::post()
                .uri("/api/calendar")
                .set_json(args)
                .insert_header(ContentType::json())
                .to_request();
            let (_, resp) = test::call_service(&app, req).await.into_parts();
            let expected = serde_json::json!({
              "5121.EG.003": {
                "events": [
                  {
                    "id": 1,
                    "room_code": "5121.EG.003",
                    "start_at": "2012-01-01T00:00:00Z",
                    "end_at": "2014-01-01T00:00:00Z",
                    "stp_title_de": "Quantenteleportation",
                    "stp_title_en": "Quantum teleportation",
                    "stp_type": "Vorlesung mit Zentralübung",
                    "entry_type": "lecture",
                    "detailed_entry_type": "Abhaltung"
                  },
                  {
                    "id": 2,
                    "room_code": "5121.EG.003",
                    "start_at": "2014-01-01T00:00:00Z",
                    "end_at": "2016-01-01T00:00:00Z",
                    "stp_title_de": "Quantenteleportation 2",
                    "stp_title_en": "Quantum teleportation 2",
                    "stp_type": "Vorlesung mit Zentralübung",
                    "entry_type": "lecture",
                    "detailed_entry_type": "Abhaltung"
                  }
                ],
                "location": {
                  "key": "5121.EG.003",
                  "name": "5121.EG.003 (Computerraum)",
                  "last_calendar_scrape_at": now,
                  "calendar_url": "https://campus.tum.de/3",
                  "type_common_name": "Serverraum",
                  "type": "room"
                }
              }
            });
            run_testcase(resp, 200, &expected.to_string()).await;
        }
        {
            // show both rooms, but a limited timeframe
            let args = Arguments {
                start_after: *TIME_2012,
                end_before: *TIME_2014,
                ids: vec!["5121.EG.003".into(), "5121.EG.001".into()],
            };
            let req = test::TestRequest::post()
                .uri("/api/calendar")
                .set_json(args)
                .insert_header(ContentType::json())
                .to_request();
            let (_, resp) = test::call_service(&app, req).await.into_parts();
            let expected = serde_json::json!({
                "5121.EG.001": {
                    "events": [],
                    "location": {
                        "calendar_url": "https://campus.tum.de/1",
                        "key": "5121.EG.001",
                        "last_calendar_scrape_at": now,
                        "name": "5121.EG.001 (Montage- und Versuchshalle)",
                        "type": "room",
                        "type_common_name": "Versuchshalle",
                    },
                },
                "5121.EG.003": {
                    "events": [
                        {
                            "detailed_entry_type": "Abhaltung",
                            "end_at": "2014-01-01T00:00:00Z",
                            "entry_type": "lecture",
                            "id": 1,
                            "room_code": "5121.EG.003",
                            "start_at": "2012-01-01T00:00:00Z",
                            "stp_title_de": "Quantenteleportation",
                            "stp_title_en": "Quantum teleportation",
                            "stp_type": "Vorlesung mit Zentralübung",
                        },
                    ],
                    "location": {
                        "calendar_url": "https://campus.tum.de/3",
                        "key": "5121.EG.003",
                        "last_calendar_scrape_at": now,
                        "name": "5121.EG.003 (Computerraum)",
                        "type": "room",
                        "type_common_name": "Serverraum",
                    },
                },
            });
            run_testcase(resp, 200, &expected.to_string()).await;
        }
    }

    async fn run_testcase(resp: HttpResponse, expected_status: u16, expected_body: &str) {
        let actual_status = resp.status().as_u16();
        let body_box = resp.into_body();
        let body_bytes = actix_web::body::to_bytes(body_box).await.unwrap();
        let body_text = String::from_utf8(body_bytes.into_iter().collect()).unwrap();
        // if the expected value cleanly deserializes into json, we should compare using this
        if let Ok(expected_value) = serde_json::from_str::<Value>(expected_body) {
            let actual_value = serde_json::from_str::<Value>(&body_text).unwrap();
            assert_eq!(actual_value, expected_value);
        } else {
            assert_eq!(body_text, expected_body);
        }
        assert_eq!(actual_status, expected_status);
    }
}
