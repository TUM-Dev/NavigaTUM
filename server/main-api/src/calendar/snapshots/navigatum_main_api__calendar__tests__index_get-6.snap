---
source: main-api/src/calendar/mod.rs
expression: actual
---
5121.EG.001:
  events: []
  location:
    calendar_url: "https://campus.tum.de/1"
    key: 5121.EG.001
    last_calendar_scrape_at: "[last_calendar_scrape_at]"
    name: 5121.EG.001 (Montage- und Versuchshalle)
    type: room
    type_common_name: Versuchshalle
5121.EG.003:
  events: []
  location:
    calendar_url: "https://campus.tum.de/3"
    key: 5121.EG.003
    last_calendar_scrape_at: "[last_calendar_scrape_at]"
    name: 5121.EG.003 (Computerraum)
    type: room
    type_common_name: Serverraum
